### PR TITLE
[MRG] ENH added automatic conversion to bv for fif if no meg

### PR DESCRIPTION
**Pull request by alexrockhill**
Thanks for contributing. If this is your first time,
make sure to read [contributing.md](https://github.com/mne-tools/mne-bids/blob/master/CONTRIBUTING.md)

PR Description
--------------

I didn't add pybv to the packages automatically installed by pip yet, so I think that still has to be done. I can look into it soon but maybe someone else knows how to add that quicker.

As is this converts a FIF file given to write_raw_bids to BV format if there are no meg channels. It uses both events_from_annotations and find_events for backward compatibility. The conversion is accurate to 1e-9 or greater when taken round trip using a data set I had.

closes

Merge checklist
---------------

Maintainer, please confirm the following before merging:

- [x] All comments resolved
- [x] This is not your own PR
- [x] All CIs are happy
- [x] PR title starts with [MRG]
- [x] [whats_new.rst](https://github.com/mne-tools/mne-bids/blob/master/doc/whats_new.rst) is updated
- [x] PR description includes phrase "closes <#issue-number>"
- [x] Commit history does not contain any merge commits

### CODE REVIEW

File: mne_bids/write.py [hunk 1]
```diff
@@ -899,21 +900,34 @@ def write_raw_bids(raw, bids_basename, output_path, events_data=None,
...
     # Copy the imaging data files
     if ext in ['.fif']:
-        n_rawfiles = len(raw.filenames)
-        if n_rawfiles > 1:
-            split_naming = 'bids'
-            raw.save(bids_fname, split_naming=split_naming, overwrite=True)
-        else:
-            # This ensures that single FIF files do not have the part param
-            raw.save(bids_fname, overwrite=True, split_naming='neuromag')
+        if pick_types(raw.info, meg=True):
```
Comment (jasmainak) [inline mne_bids/write.py]: you can just do:

`if 'meg' in raw`

or reuse the `kind` variable above

File: mne_bids/write.py [hunk 2]
```diff
@@ -16,14 +16,15 @@
...
-from mne import Epochs
+from mne import Epochs, pick_types, find_events, events_from_annotations
 from mne.io.constants import FIFF
 from mne.io.pick import channel_type
 from mne.io import BaseRaw
 from mne.channels.channels import _unit2human
 from mne.utils import check_version, has_nibabel
 from mne.transforms import _ensure_trans, apply_trans
 
+from pybv import write_brainvision
```
Comment (jasmainak) [inline mne_bids/write.py]: I would not do a hard dependency. Only if you are attempting to write EEG data would I import this

File: mne_bids/write.py (hunk 1 above)
Comment (jasmainak) [reply mne_bids/write.py]: I think reusing the `kind` variable is the way to go

**Comment by jasmainak** [comment]
@alexrockhill can you add a test?

**Comment by jasmainak** [comment]
@alexrockhill any progress here? Let me know if you need a hand and want to pair program for this PR ...

**Comment by alexrockhill** [comment]
Hey @jasmainak, I incorporated your suggestions so that any file type given would be written in BIDS even if it were meg and not in the compatible types for instance but more applicably eeg files that are not in the compatible types are converted to BrainVision format. If you have a second, I'm getting the following rather cryptic error in my testing and I thought you might know a better way to refactor the test or avoid the error:

Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "/autofs/space/karima_001/users/alex/software/anaconda3/lib/python3.6/site-packages/mne_bids-0.2.dev0-py3.6.egg/mne_bids/write.py", line 885, in write_raw_bids
    _events_tsv(events, raw, events_fname, event_id, overwrite, verbose)
  File "/autofs/space/karima_001/users/alex/software/anaconda3/lib/python3.6/site-packages/mne_bids-0.2.dev0-py3.6.egg/mne_bids/write.py", line 172, in _events_tsv
    trial_type_map = {v: k for k, v in trial_type.items()}
AttributeError: 'str' object has no attribute 'items'

File: mne_bids/tests/test_write.py [hunk 3]
```diff
@@ -92,6 +92,17 @@ def test_fif():
     cmd = bids_validator_exe + [output_path]
     run_subprocess(cmd, shell=shell)
 
+    #try with eeg data only (conversion to bv)
+    output_path = _TempDir()
+    raw = mne.io.read_raw_fif(raw_fname)
+    raw.load_data()
+    raw = raw.pick_types(meg=False, eeg=True, stim=True, eog=True, ecg=True)
+    raw.save(op.join(data_path, 'test.fif'))
```
Comment (jasmainak) [inline mne_bids/tests/test_write.py]: you should use `overwrite=True` here

File: mne_bids/write.py [hunk 4]
```diff
@@ -898,23 +897,41 @@ def write_raw_bids(raw, bids_basename, output_path, events_data=None,
...
+            else:
+                # This ensures that single FIF files do not have the part param
+                raw.save(bids_fname, overwrite=True, split_naming='neuromag')
+        else:
+            if verbose:
+                print('Converting data files to BrainVision format')
+            try:
+                from pybv import write_brainvision
+            except ImportError:
+                raise ImportError('pybv required in order to convert FIF to BV format')
```
Comment (jasmainak) [inline mne_bids/write.py]: you should use a pep8 linter on your editor.

**Comment by jasmainak** [comment]
@alexrockhill what version of `pybv` are you using?

I get the following when I try to run the tests:

```py
E               ValueError: events must be an (n_events x 2) ndarray or None'
```

It seems like a `pybv` issue that @sappelhoff may know about

**Comment by jasmainak** [comment]
@alexrockhill one tip: just add the `--pdb` flag when you run pytests and you'll drop right into debug mode when you fail, so you can easily figure out what is the issue

**Comment by sappelhoff** [comment]
pybv expects the following events:

> events (ndarray, shape (n_events, 2)) – Events to write in the marker file. This array has two columns. The first column is the zero-based index of each event (corresponding to the “time” dimension of the data array). The second column is a number associated with the “type” of event.

see the [API](https://pybv.readthedocs.io/en/latest/generated/pybv.write_brainvision.html#pybv.write_brainvision)

What's the trouble?

**Comment by jasmainak** [comment]
@alexrockhill you need to fix this according to the comment above

**Comment by alexrockhill** [comment]
@jasmainak Sorry forgot to add that line, I should have had a test :)

**Comment by jasmainak** [comment]
@alexrockhill you need to rebase

File: mne_bids/write.py [hunk 5]
```diff
@@ -893,23 +899,42 @@ def write_raw_bids(raw, bids_basename, output_path, events_data=None,
...
-    if ext not in ALLOWED_EXTENSIONS:
-        raise ValueError('ext must be in %s, got %s'
-                         % (''.join(ALLOWED_EXTENSIONS), ext))
-
     # Copy the imaging data files
-    if ext in ['.fif']:
+    if ext not in ALLOWED_EXTENSIONS[kind]:
+        if kind == 'meg':
+            if verbose:
+                print('Converting data files to Fractal Image Format (FIF)')
```
Comment (teonbrooks) [inline mne_bids/write.py]: FIF in this case stands for [Functional Image File](http://www.nmr.mgh.harvard.edu/CFNT/sofMNE.html)

Comment (alexrockhill) [reply mne_bids/write.py]: Oops :)

File: mne_bids/write.py [hunk 6]
```diff
@@ -895,23 +901,42 @@ def write_raw_bids(raw, bids_basename, output_path, events_data=None,
...
+                split_naming = 'bids'
+                raw.save(bids_fname, split_naming=split_naming, overwrite=True)
+            else:
+                # This ensures that single FIF files do not have the part param
+                raw.save(bids_fname, overwrite=True, split_naming='neuromag')
+        else:
+            if verbose:
+                print('Converting data files to BrainVision format')
+            try:
+                from pybv import write_brainvision
```
Comment (sappelhoff) [inline mne_bids/write.py]: if pybv is used, we should add a note to the README, which currently says:
```
For full functionality of mne-bids, you will also need to pip install the following packages:

    nibabel

```

then the list of "nibabel" can be extended to include pybv

also, pybv needs to be added to the [`environment.yml`](https://github.com/mne-tools/mne-bids/blob/master/environment.yml) file

File: mne_bids/write.py [hunk 7]
```diff
@@ -895,23 +901,42 @@ def write_raw_bids(raw, bids_basename, output_path, events_data=None,
...
-    if ext not in ALLOWED_EXTENSIONS:
-        raise ValueError('ext must be in %s, got %s'
-                         % (''.join(ALLOWED_EXTENSIONS), ext))
-
     # Copy the imaging data files
-    if ext in ['.fif']:
+    if ext not in ALLOWED_EXTENSIONS[kind]:
+        if kind == 'meg':
+            if verbose:
+                print('Converting data files to Functional Image Format (FIF)')
```
Comment (jasmainak) [inline mne_bids/write.py]: use `warnings.warn`

File: mne_bids/write.py [hunk 8]
```diff
@@ -898,12 +899,33 @@ def write_raw_bids(raw, bids_basename, output_path, events_data=None,
...
+            else:
+                # This ensures that single FIF files do not have the part param
+                raw.save(bids_fname, overwrite=True, split_naming='neuromag')
+        elif kind == 'eeg':
+            if verbose:
+                warn('Converting data files to BrainVision format')
+            try:
+                from pybv import write_brainvision
+            except ImportError:
+                raise ImportError('This function requoires pybv' +
```
Comment (agramfort) [inline mne_bids/write.py]:
```suggestion
                raise ImportError('This function requires pybv '
```

File: README.rst [hunk 9]
```diff
@@ -90,6 +90,7 @@ For full functionality of ``mne-bids``, you will also need to ``pip install``
 the following packages:
 
 - ``nibabel``, for interacting with MRI data
+- ``pybv``, for converting FIF format EEG data to BV to be valid according to EEG BIDS specifications
```
Comment (agramfort) [inline README.rst]:
```suggestion
- ``pybv``, to convert EEG data to BrainVision if input format is not be valid according to EEG BIDS specifications
```

File: mne_bids/write.py [hunk 10]
```diff
@@ -898,12 +899,33 @@ def write_raw_bids(raw, bids_basename, output_path, events_data=None,
...
+        if kind == 'meg':
+            if verbose:
+                warn('Converting data files to Functional Image Format (FIF)')
+            n_rawfiles = len(raw.filenames)
+            if n_rawfiles > 1:
+                split_naming = 'bids'
+                raw.save(bids_fname, split_naming=split_naming, overwrite=True)
+            else:
+                # This ensures that single FIF files do not have the part param
+                raw.save(bids_fname, overwrite=True, split_naming='neuromag')
```
Comment (agramfort) [inline mne_bids/write.py]:
```suggestion
                raw.save(bids_fname, split_naming='neuromag', overwrite=True)
```

**Comment by jasmainak** [comment]
can you update whats_new page @alexrockhill ?

**Comment by jasmainak** [comment]
can you also add the keyword "close <issue-number>" in the description of PR and set the title to [MRG] when you are ready?

File: doc/whats_new.rst [hunk 11]
```diff
@@ -22,6 +22,8 @@ Current
 Changelog
 ~~~~~~~~~
 
+- Added automatic conversion of FIF to BV with warning for EEG only data and conversion to FIF for meg non-FIF data, by `Alex Rockhill` _ (`#237 <https://github.com/mne-tools/mne-bids/pull/237>`_)
```
Comment (agramfort) [inline doc/whats_new.rst]: please avoid acronyms BV -> BrainVision
```suggestion
- Added automatic conversion of FIF to BrainVision format with warning for EEG only data and conversion to FIF for meg non-FIF data, by `Alex Rockhill` _ (`#237 <https://github.com/mne-tools/mne-bids/pull/237>`_)
```

File: README.rst [hunk 12]
```diff
@@ -90,6 +90,7 @@ For full functionality of ``mne-bids``, you will also need to ``pip install``
 the following packages:
 
 - ``nibabel``, for interacting with MRI data
+- ``pybv``, to convert EEG data to BrainVision if input format is not be valid according to EEG BIDS specifications
```
Comment (agramfort) [inline README.rst]:
```suggestion
- ``pybv``, to convert EEG data to BrainVision if input format is not valid according to EEG BIDS specifications
```

File: mne_bids/write.py [hunk 13]
```diff
@@ -898,12 +899,33 @@ def write_raw_bids(raw, bids_basename, output_path, events_data=None,
...
+                                  'in order to convert FIF to BV format.')
+            events, event_id = events_from_annotations(raw)
+            write_brainvision(raw.get_data(), raw.info['sfreq'],
+                              raw.ch_names,
+                              op.splitext(op.basename(bids_fname))[0],
+                              op.dirname(bids_fname), events[:, [0, 2]],
+                              resolution=1e-6)
     # Copy the imaging data files
-    if ext in ['.fif']:
+    elif ext in ['.fif']:
```
Comment (agramfort) [inline mne_bids/write.py]:
```suggestion
    elif ext == '.fif':
```

File: mne_bids/write.py [hunk 14]
```diff
@@ -898,12 +899,33 @@ def write_raw_bids(raw, bids_basename, output_path, events_data=None,
...
-                         % (''.join(ALLOWED_EXTENSIONS), ext))
-
+    if ext not in ALLOWED_EXTENSIONS[kind]:
+        if kind == 'meg':
+            if verbose:
+                warn('Converting data files to Functional Image Format (FIF)')
+            n_rawfiles = len(raw.filenames)
+            if n_rawfiles > 1:
+                split_naming = 'bids'
+                raw.save(bids_fname, split_naming=split_naming, overwrite=True)
```
Comment (agramfort) [inline mne_bids/write.py]: these lines are not covered by tests apparently.

You should avoid the code duplication with the block

```
        n_rawfiles = len(raw.filenames)
        if n_rawfiles > 1:
            split_naming = 'bids'
            raw.save(bids_fname, split_naming=split_naming, overwrite=True)
        else:
            # This ensures that single FIF files do not have the part param
            raw.save(bids_fname, overwrite=True, split_naming='neuromag')
```

that you have below. See how you can do this by rethinking the branching logic.

**Comment by jasmainak** [comment]
Can you rebase the branch?

File: mne_bids/write.py [hunk 15]
```diff
@@ -898,19 +899,34 @@ def write_raw_bids(raw, bids_basename, output_path, events_data=None,
...
 
-    if ext not in ALLOWED_EXTENSIONS:
-        raise ValueError('ext must be in %s, got %s'
-                         % (''.join(ALLOWED_EXTENSIONS), ext))
-
+    convert = ext not in ALLOWED_EXTENSIONS[kind]
     # Copy the imaging data files
-    if ext in ['.fif']:
+    if convert and kind == 'ieeg':
+        raise ValueError('Conversion for ieeg has not been implemented yet.')
```
Comment (jasmainak) [inline mne_bids/write.py]: what stops you from doing this? Doesn't iEEG format also support Brainvision?

File: mne_bids/write.py [hunk 16]
```diff
@@ -898,19 +899,32 @@ def write_raw_bids(raw, bids_basename, output_path, events_data=None,
...
         print('Copying data files to %s' % bids_fname)
 
-    if ext not in ALLOWED_EXTENSIONS:
-        raise ValueError('ext must be in %s, got %s'
-                         % (''.join(ALLOWED_EXTENSIONS), ext))
-
+    convert = ext not in ALLOWED_EXTENSIONS[kind]
     # Copy the imaging data files
-    if ext in ['.fif']:
+    if convert and (kind == 'eeg' or kind == 'ieeg'):
```
Comment (agramfort) [inline mne_bids/write.py]:
```suggestion
    if convert and (kind in ['eeg', 'ieeg']):
```

File: doc/whats_new.rst [hunk 17]
```diff
@@ -22,6 +22,8 @@ Current
 Changelog
 ~~~~~~~~~
 
+- Added automatic conversion of FIF to BrainVision format with warning for EEG only data and conversion to FIF for meg non-FIF data, by `Alex Rockhill` _ (`#237 <https://github.com/mne-tools/mne-bids/pull/237>`_)
```
Comment (agramfort) [inline doc/whats_new.rst]:
```suggestion
- Added automatic conversion to BrainVision format for EEG only data and to FIF for meg (non-FIF) data, by `Alex Rockhill` _ (`#237 <https://github.com/mne-tools/mne-bids/pull/237>`_)
```

Comment (jasmainak) [reply doc/whats_new.rst]: Actually, this will not render correctly because you don't have your name down below with link to website

Comment (alexrockhill) [reply doc/whats_new.rst]: Fixed

**Comment by jasmainak** [comment]
this PR now has merge conflicts @alexrockhill

**Comment by alexrockhill** [comment]
Not sure why the appveyor is failing but I think this is ready to merge with @sappelhoff's approval. The appveyor passed on a previous iteration with minor changes, not sure what's going on there.

File: mne_bids/write.py [hunk 18]
```diff
@@ -777,9 +777,10 @@ def write_raw_bids(raw, bids_basename, output_path, events_data=None,
                    event_id=None, overwrite=False, verbose=True):
     """Walk over a folder of files and create BIDS compatible folder.
 
-    .. warning:: The original files are simply copied over. This function
-                 cannot convert modify data files from one format to another.
-                 Modification of the original data files is not allowed.
+    .. warning:: The original files are simply copied over if the original
+                 file format is BIDS-supported for that modality. Otherwise,
+                 this function will convert to a BIDS-supported file format
```
Comment (jasmainak) [inline mne_bids/write.py]: Might be nice to be more specific. Say that you convert to brainvision format for EEG (and iEEG?) data.

**Comment by sappelhoff** [comment]
re: appveyor, see: https://github.com/mne-tools/mne-bids/pull/275

**Comment by jasmainak** [comment]
@sappelhoff you beat me to it :)

This PR should be good now as I rebased it against master

**Comment by alexrockhill** [comment]
Ok I fixed @jasmainak's comment about specifying which formats are converted to automatically in the doc. (When run, there is a specific warning depending on the input data type already). I think should be okay to merge except it looks like there's still issues with Appveyor:

```git : Cloning into 'mne-python'...
At line:2 char:1
+ git clone --depth 1 https://github.com/mne-tools/mne-python.git -b $e ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (Cloning into 'mne-python'...:String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError
 
running develop
running egg_info

**Comment by alexrockhill** [comment]
> Ok I fixed @jasmainak's comment about specifying which formats are converted to automatically in the doc. (When run, there is a specific warning depending on the input data type already). I think should be okay to merge except it looks like there's still issues with Appveyor:
> 
> ```
> At line:2 char:1
> + git clone --depth 1 https://github.com/mne-tools/mne-python.git -b $e ...
> + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
>     + CategoryInfo          : NotSpecified: (Cloning into 'mne-python'...:String) [], RemoteException
>     + FullyQualifiedErrorId : NativeCommandError
>  
> running develop
> running egg_info
> ```

Sorry, @sappelhoff would you happen to have a second to look at this? If you don't have time I can but since you were just fixing Appveyor it might be easier for you.

**Comment by alexrockhill** [comment]
Maybe this is helpful: https://help.appveyor.com/discussions/problems/3788-appveyor-failing-builds-due-to-bitbucket-message

**Comment by alexrockhill** [comment]
Yay fixed appveyor! @sappelhoff @jasmainak please let me know if I did it wrong and that's not how appveyor should be written. 

Should be okay to merge, I think.

After this is merged, maybe @bloyl will help me look over putting anonymize in mne-bids. These are related because I think we'll have to save to FIF then convert to BV for EEG and iEEG if we want to use anonymize_info.

**Comment by alexrockhill** [comment]
One sec, I'll add a couple tests first to increase coverage actually

**Comment by alexrockhill** [comment]
Ok so if @jasmainak or anyone has any suggestions I'm running into issues because I don't think there is a use case for an MEG format that isn't supported by bids so it doesn't make sense to create one to test meg -> FIF. I think I did the test pybv correctly but let me know if that's wrong...

I think it's worthwhile having meg conversion in there for symmetry and potentially future use but I guess it could be taken out. Input appreciated.

File: mne_bids/write.py [hunk 19]
```diff
@@ -965,21 +967,36 @@ def write_raw_bids(raw, bids_basename, output_path, events_data=None,
...
+                              'in order to convert FIF to BV format.')
+        events, event_id = events_from_annotations(raw)
+        write_brainvision(raw.get_data(), raw.info['sfreq'],
+                          raw.ch_names,
+                          op.splitext(op.basename(bids_fname))[0],
+                          op.dirname(bids_fname), events[:, [0, 2]],
+                          resolution=1e-6)
+    elif ext == '.fif' or (convert and kind == 'meg'):
+        ''' I don't think there are any current uses for this
+            so it doesn't make sense to write tests.'''
```
Comment (jasmainak) [inline mne_bids/write.py]: okay now I'm confused by the logic. Can you explain it to me in plain English?

File: mne_bids/write.py [hunk 20]
```diff
@@ -965,21 +967,36 @@ def write_raw_bids(raw, bids_basename, output_path, events_data=None,
...
 
     if verbose:
-        print('Copying data files to %s' % bids_fname)
-
-    if ext not in ALLOWED_EXTENSIONS:
-        raise ValueError('ext must be in %s, got %s'
-                         % (''.join(ALLOWED_EXTENSIONS), ext))
+        print('Copying data files to %s' % op.splitext(bids_fname)[0])
 
+    convert = ext not in ALLOWED_EXTENSIONS[kind]
```
Comment (jasmainak) [inline mne_bids/write.py]: what happens if `ext` is 'blahblah' ?

**Comment by alexrockhill** [comment]
Ok, I think this is okay to merge.

**Comment by alexrockhill** [comment]
Actually I could add one that just changed the name of the meg extension to .xxx or something, that could test the conversion

File: mne_bids/write.py [hunk 21]
```diff
@@ -965,21 +967,33 @@ def write_raw_bids(raw, bids_basename, output_path, events_data=None,
...
+        requires_version('pybv', '0.2.0')
+        from pybv import write_brainvision
+        events, event_id = events_from_annotations(raw)
+        write_brainvision(raw.get_data(), raw.info['sfreq'],
+                          raw.ch_names,
+                          op.splitext(op.basename(bids_fname))[0],
+                          op.dirname(bids_fname), events[:, [0, 2]],
+                          resolution=1e-6)
+    elif ext == '.fif' or (convert and kind == 'meg'):
+        ''' Until there are meg formats not supported by bids
```
Comment (jasmainak) [inline mne_bids/write.py]:
```suggestion
```

File: mne_bids/write.py [hunk 22]
```diff
@@ -965,21 +967,33 @@ def write_raw_bids(raw, bids_basename, output_path, events_data=None,
...
+            this will not be used.'''
```
Comment (jasmainak) [inline mne_bids/write.py]:
```suggestion
```

**Comment by jasmainak** [comment]
> Actually I could add one that just changed the name of the meg extension to .xxx or something, that could test the conversion

that's a fantastic idea if you could make that work

**Comment by alexrockhill** [comment]
Ok, I have spent too much time on writing a test for meg conversion, it can't be done easily:

1) If you use meg data and convert it to a different format just by changing the extension there isn't a reader.

2) If you use eeg data, you cannot set_channel_type to grad or mag because it's not supported.

Therefore until there is a use case I propose not testing this.

**Comment by alexrockhill** [comment]
Ok, I have spent too much time on writing a test for meg conversion, it can't be done easily:

1) If you use meg data and convert it to a different format just by changing the extension there isn't a reader.

2) If you use eeg data, you cannot set_channel_type to grad or mag because it's not supported.

Therefore until there is a use case I propose not testing this. Sorry for vacillating on whether this is done, I am good with this being merged now.

File: mne_bids/write.py [hunk 23]
```diff
@@ -965,21 +967,31 @@ def write_raw_bids(raw, bids_basename, output_path, events_data=None,
...
-                         % (''.join(ALLOWED_EXTENSIONS), ext))
+        print('Copying data files to %s' % op.splitext(bids_fname)[0])
 
+    convert = ext not in ALLOWED_EXTENSIONS[kind]
     # Copy the imaging data files
-    if ext in ['.fif']:
+    if convert and (kind == 'eeg' or kind == 'ieeg'):
+        if verbose:
+            warn('Converting data files to BrainVision format')
+        requires_version('pybv', '0.2.0')
```
Comment (jasmainak) [inline mne_bids/write.py]: I think you should use `check_version` here. See how we do it in MNE: https://github.com/mne-tools/mne-python/blob/9e588fca0007e8bda43c8a1349d679aaa71c0768/mne/preprocessing/ica.py#L330

File: README.rst [hunk 24]
```diff
@@ -90,6 +90,7 @@ For full functionality of ``mne-bids``, you will also need to ``pip install``
 the following packages:
 
 - ``nibabel``, for interacting with MRI data
+- ``pybv``, to convert EEG data to BrainVision if input format is not valid according to EEG BIDS specifications
```
Comment (jasmainak) [inline README.rst]: you'll need to rebase add a line to the list of dependencies

**Comment by jasmainak** [comment]
> If you use meg data and convert it to a different format just by changing the extension there isn't a reader.

So is there a practical use case to support conversion for MEG data? If not, let's not support it.

File: mne_bids/write.py [hunk 25]
```diff
@@ -965,21 +967,31 @@ def write_raw_bids(raw, bids_basename, output_path, events_data=None,
...
+        from pybv import write_brainvision
+        events, event_id = events_from_annotations(raw)
+        write_brainvision(raw.get_data(), raw.info['sfreq'],
+                          raw.ch_names,
+                          op.splitext(op.basename(bids_fname))[0],
+                          op.dirname(bids_fname), events[:, [0, 2]],
+                          resolution=1e-6)
+    elif ext == '.fif' or (convert and kind == 'meg'):
```
Comment (jasmainak) [inline mne_bids/write.py]:
```suggestion
    elif ext == '.fif':
```

File: mne_bids/write.py [hunk 26]
```diff
@@ -965,21 +967,31 @@ def write_raw_bids(raw, bids_basename, output_path, events_data=None,
...
+        if convert and verbose:
```
Comment (jasmainak) [inline mne_bids/write.py]:
```suggestion
```

File: mne_bids/write.py [hunk 27]
```diff
@@ -965,21 +967,31 @@ def write_raw_bids(raw, bids_basename, output_path, events_data=None,
...
+            warn('Converting data files to Functional Image Format (FIF)')
```
Comment (jasmainak) [inline mne_bids/write.py]:
```suggestion
```

File: mne_bids/write.py [hunk 28]
```diff
@@ -965,21 +967,31 @@ def write_raw_bids(raw, bids_basename, output_path, events_data=None,
...
-
-    if ext not in ALLOWED_EXTENSIONS:
-        raise ValueError('ext must be in %s, got %s'
-                         % (''.join(ALLOWED_EXTENSIONS), ext))
+        print('Copying data files to %s' % op.splitext(bids_fname)[0])
 
+    convert = ext not in ALLOWED_EXTENSIONS[kind]
     # Copy the imaging data files
-    if ext in ['.fif']:
+    if convert and (kind == 'eeg' or kind == 'ieeg'):
```
Comment (jasmainak) [inline mne_bids/write.py]:
```suggestion
    if convert:
        if kind == 'meg':
            raise ValueError('Got file extension %s for MEG data, expected one of %s' % ALLOWED_EXT['meg'])
```
something along these lines

File: mne_bids/write.py [hunk 29]
```diff
@@ -965,21 +967,32 @@ def write_raw_bids(raw, bids_basename, output_path, events_data=None,
...
+    convert = ext not in ALLOWED_EXTENSIONS[kind]
     # Copy the imaging data files
-    if ext in ['.fif']:
+    if convert:
+        if kind == 'meg':
+            raise ValueError('Got file extension %s for MEG data, ' +
+                             'expected one of %s' % ALLOWED_EXTENSIONS['meg'])
+        if verbose:
+            warn('Converting data files to BrainVision format')
+        check_version('pybv', '0.2')
```
Comment (jasmainak) [inline mne_bids/write.py]: I don't think you're using this correctly. Not sure what you want to do with the code here but `check_version` just returns a boolean and does not throw an error if `pybv` is not installed.

**Comment by alexrockhill** [comment]
Good by me

File: mne_bids/write.py [hunk 30]
```diff
@@ -965,21 +967,33 @@ def write_raw_bids(raw, bids_basename, output_path, events_data=None,
...
     # Copy the imaging data files
-    if ext in ['.fif']:
+    if convert:
+        if kind == 'meg':
+            raise ValueError('Got file extension %s for MEG data, ' +
+                             'expected one of %s' % ALLOWED_EXTENSIONS['meg'])
+        if verbose:
+            warn('Converting data files to BrainVision format')
+        if not check_version('pybv', '0.2'):
+            raise ImportError('pybv is required for this conversion')
```
Comment (jasmainak) [inline mne_bids/write.py]: Sorry last (I promise!) comment. Can you make this a more informative error?

```suggestion
            raise ImportError('pybv >=0.2.0 is required for converting %s files to Brainvision format' % ext)
```

**Comment by jasmainak** [comment]
Merged, thanks @alexrockhill for the contribution, and for being patient with the reviews :)
